### PR TITLE
[PEPC-Boost][2] Block Assembler Implementation

### PR DESCRIPTION
**Pull request by bharath-123**
## 📝 Summary

This PR contains an implementation of the Block Assembler RPC.

The Block Assembler RPC is called by the PEPC-Boost relay. When the builder submits a block to the PEPC-Boost relayer, the relayer checks the existing ToB tx cache for the particular slot and parent hash. If it finds the ToB txs, the sum of the ToB tx value and RoB tx value is the highest seen so far. We then call the Block Assembler RPC API, sending the ToB txs, and the Builder submit request as parameters. The block assembler assembles the block with the ToB txs at the top and the RoB txs in the rest of the block. If successful, it returns the final execution payload to the relayer which it stores as the final block.

### CODE REVIEW

File: eth/block-validation/api.go [hunk 1]
```diff
@@ -344,3 +346,89 @@ func (b *BlockAssemblerRequest) UnmarshalJSON(data []byte) error {
 
 	return nil
 }
+
+func (api *BlockValidationAPI) BlockAssembler(params *BlockAssemblerRequest) (*capella.ExecutionPayload, error) {
+	log.Info("BlockAssembler", "tobTxs", len(params.TobTxs.Transactions), "robPayload", params.RobPayload)
+	transactionBytes := make([][]byte, len(params.TobTxs.Transactions))
+	for i, txHexBytes := range params.TobTxs.Transactions {
```
Comment (michaelneuder) [inline eth/block-validation/api.go:353]: why is this copy needed ?

File: eth/block-validation/api.go [hunk 2]
```diff
@@ -344,3 +346,89 @@ func (b *BlockAssemblerRequest) UnmarshalJSON(data []byte) error {
...
+	// TODO - check for gas limits
+	// TODO - support for payouts
+
+	// TODO - if there are no TOB txs then we can just simulate the block rather then re-assembling it.
+	// These are TODOs for now to simplify integration testing
+
+	// check if there are any duplicate txs
+	// we can error out if there is a nonce gap
+	// TODO - don't error out, but drop the duplicate tx in the ROB block
+	seenTxMap := make(map[common.Hash]struct{})
```
Comment (michaelneuder) [inline eth/block-validation/api.go:377]: why is the type of the map an empty struct? bool seems clearer

File: eth/block-validation/api.go [hunk 3]
```diff
@@ -344,3 +346,89 @@ func (b *BlockAssemblerRequest) UnmarshalJSON(data []byte) error {
...
+	for _, tx := range robBlock.Transactions() {
+		// if we see nonce re-use in TOB vs ROB then drop txs the txs in ROB
+		if _, ok := seenTxMap[tx.Hash()]; ok {
+			return nil, errors.New("duplicate tx")
+		}
+		seenTxMap[tx.Hash()] = struct{}{}
+	}
+
+	withdrawals := make(types.Withdrawals, len(params.RobPayload.ExecutionPayload.Withdrawals))
+	for i, withdrawal := range params.RobPayload.ExecutionPayload.Withdrawals {
```
Comment (michaelneuder) [inline eth/block-validation/api.go:394]: confused why this copy is needed ?

File: eth/block-validation/api.go [hunk 4]
```diff
@@ -344,3 +346,89 @@ func (b *BlockAssemblerRequest) UnmarshalJSON(data []byte) error {
...
+	resolvedBlock := block.ResolveFull()
+	if resolvedBlock == nil {
+		return nil, errors.New("unable to resolve block")
+	}
+	if resolvedBlock.ExecutionPayload == nil {
+		return nil, errors.New("nil execution payload")
+	}
+
+	finalPayload, err := engine.ExecutableDataToCapellaExecutionPayload(resolvedBlock.ExecutionPayload)
+
```
Comment (michaelneuder) [inline eth/block-validation/api.go:435]: need to check this err?

File: miner/multi_worker.go [hunk 5]
```diff
@@ -86,14 +86,60 @@ func (w *multiWorker) disablePreseal() {
 	}
 }
 
+// buildPayload builds the payload according to the provided parameters.
+func (w *multiWorker) payloadAssembler(args *BuildPayloadArgs) (*Payload, error) {
```
Comment (michaelneuder) [inline miner/multi_worker.go:90]: curious if this is copied from somewhere?

File: miner/worker.go [hunk 6]
```diff
@@ -1095,6 +1102,95 @@ func (w *worker) commitBundle(env *environment, txs types.Transactions, interrup
 	return nil
 }
 
+func (w *worker) commitTransactionWrapper(env *environment, interrupt *int32, tx *types.Transaction, coalescedLogs []*types.Log) error {
```
Comment (michaelneuder) [inline miner/worker.go:1105]: curious if this is copied from somewhere ?

**Review by michaelneuder — COMMENTED** [review]
LGTM overall to me. a few small comments.

**Comment by michaelneuder** [comment]
high level feedback: could you do a changelog style summary on the PRs? e.g., just super high level description of the files and how they fit together ? would really help visualize the flow. also just an example lifecycle of a ToB txn and a RoB txn would be insanely helpful

File: eth/block-validation/api.go (hunk 2 above)
Comment (bharath-123) [reply eth/block-validation/api.go:377]: true it is cleaner. I have seen this in the current code with mempoolHashes but bool just makes more sense yes

File: eth/block-validation/api.go (hunk 4 above)
Comment (bharath-123) [reply eth/block-validation/api.go:435]: you are right, will do that

File: miner/multi_worker.go (hunk 5 above)
Comment (bharath-123) [reply miner/multi_worker.go:90]: it is built on the template of `buildPayload` dude. the major changes are in `sealingBlock`

File: miner/worker.go (hunk 6 above)
Comment (bharath-123) [reply miner/worker.go:1105]: The wrapper is not copied from somewhere. just abstracted out redundant code from `commitAssemblerTransactions`

**Comment by bharath-123** [comment]
> high level feedback: could you do a changelog style summary on the PRs? e.g., just super high level description of the files and how they fit together ? would really help visualize the flow. also just an example lifecycle of a ToB txn and a RoB txn would be insanely helpful

Yes I will do that! Will update you once done!

File: eth/block-validation/api.go (hunk 3 above)
Comment (bharath-123) [reply eth/block-validation/api.go:394]: to convert from `[]*Withdrawal` to `types.Withdrawal` type. Will check if I can avoid it

File: eth/block-validation/api.go (hunk 1 above)
Comment (bharath-123) [reply eth/block-validation/api.go:353]: type conversions from `[]bellatrix.Transaction` to `[][]byte`. Lot of opaque types here dude. Need to check if there is  cleaner way to do it.

**Pull request merged**